Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt

```diff
@@ -46,16 +53,38 @@ class DartToolingDaemonService private constructor(private val project: Project)
     result.exePath = FileUtil.toSystemDependentName(DartSdkUtil.getDartExePath(sdk))
     result.charset = StandardCharsets.UTF_8
     result.addParameter("tooling-daemon")
+    result.addParameter("--machine")
 
     dtdProcessHandler = ColoredProcessHandler(result)
     dtdProcessHandler.addProcessListener(object : ProcessListener {
       override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
         val text = event.text.trim()
-        if (text.startsWith(STARTUP_MESSAGE_PREFIX)) {
-          dtdProcessHandler.removeProcessListener(this)
-          val address: String = text.substring(STARTUP_MESSAGE_PREFIX.length)
-          connectToDtdWebSocket(address)
+
+        // The first line of text is the command issued, which can be ignored.
+        if (text.endsWith("dart tooling-daemon --machine")) {
+          return
+        }
+
+        val json: JsonObject
+        try {
+          json = JsonParser.parseString(text) as JsonObject
+        }
+        catch (e: Exception) {
+          dtdInfoFuture.completeExceptionally(Exception("Parse message failed: $text $e"))
+          return
         }
+
+        val details = json["tooling_daemon_details"]?.asJsonObject
+        if (details == null) {
+          dtdInfoFuture.completeExceptionally(Exception("No tooling daemon details found"))
+          return
+        }
+
+        val uri = details["uri"].asString
+        val secret = details["trusted_client_secret"].asString
+        dtdProcessHandler.removeProcessListener(this)
+        dtdInfoFuture.complete(DtdInfo(uri, secret))
+        connectToDtdWebSocket(uri)
```

Comment:
In my experiments, `uri` looks like `ws://127.0.0.1:65323/`. And the `connectToDtdWebSocket()` implementation is
```
private fun connectToDtdWebSocket(address: String) {
    webSocket = WebSocket(URI("ws://$address/ws"))
  ...
```
So the `URI` passed to `WebSocket()` constructor is like `ws://ws://127.0.0.1:65323//ws`

Should I change `connectToDtdWebSocket()` implementation to 
```
    webSocket = WebSocket(URI("${address}ws"))
```
?

In this example I removed the `ws://` prefix and the slash after `${address}`. Otherwize, there were two schema prefixes and two slashes in the end. Is the proposed change safe or we should introduce more complex string manipulation logic?

Reply:
Oops, sorry about that! I know I ran into this somewhere and didn't realize I changed it in the wrong PR.